### PR TITLE
Add TFVisionEncoderDecoderModel

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

To make Vision-Encoder-Text-Decoder family complete by adding `TFVisionEncoderDecoderModel`.

To complete this PR, it requires to wait #13778 being merged to master (then rebase)
(And if we want to include a real integration test using the recent image-captioning ViT + GPT2 model, need to wait #14038 too)

### CODE REVIEW

File: src/transformers/generation_tf_utils.py
Comment (sgugger) [inline src/transformers/generation_tf_utils.py:632]: Should be cleaned up?

File: src/transformers/generation_tf_utils.py
Comment (sgugger) [inline src/transformers/generation_tf_utils.py:698]: Act on the TODO and remove the comment?

File: src/transformers/generation_tf_utils.py
Comment (sgugger) [inline src/transformers/generation_tf_utils.py:714]: Reformulate the comment to just explain why there is a `len(shape_list(input_ids)) == 2` here.

File: src/transformers/models/encoder_decoder/modeling_tf_encoder_decoder.py
Comment (sgugger) [inline src/transformers/models/encoder_decoder/modeling_tf_encoder_decoder.py:222]: Why is this part removed?

File: src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py
Comment (sgugger) [inline src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:545]: There is a function for this:
```suggestion
            encoder_processing_inputs.update(kwargs_encoder)
```

File: src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py
Comment (sgugger) [inline src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:597]:
```suggestion
        decoder_processing_inputs.update(kwargs_decoder)
```

File: src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py
Comment (sgugger) [inline src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:662]: This fits in one line:
```suggestion
    def prepare_inputs_for_generation(self, decoder_input_ids, past, use_cache=None, **kwargs):
```

File: tests/test_modeling_tf_vision_encoder_decoder.py
Comment (sgugger) [inline tests/test_modeling_tf_vision_encoder_decoder.py:228]:
```suggestion
        self.assertIn("loss", outputs_encoder_decoder)
```

File: tests/test_modeling_tf_vision_encoder_decoder.py
Comment (sgugger) [inline tests/test_modeling_tf_vision_encoder_decoder.py:522]:
```suggestion
            after_outputs = model_1(pixel_values=pixel_values, decoder_input_ids=decoder_input_ids)
```

**Review by sgugger — APPROVED** [review]
Thanks for adding this model!

File: src/transformers/models/encoder_decoder/modeling_tf_encoder_decoder.py
Comment (ydshieh) [reply src/transformers/models/encoder_decoder/modeling_tf_encoder_decoder.py:222]: `TFEncoderDecoderModel.call()` doesn't have `encoder_hidden_states` parameter, but `encoder_outputs`.
Moreover, `encoder_hidden_states` is always passed to the decoder with `encoder_hidden_states = encoder_outputs[0]`. Therefore, there is no need to add `encoder_hidden_states` in `dummy_inputs`.

(I don't remember why I did that before, probably in some intermediate commits, it was required)

File: src/transformers/generation_tf_utils.py
Comment (ydshieh) [reply src/transformers/generation_tf_utils.py]: Done

**Comment by ydshieh** [comment]
@sgugger, thank you for your review, I have addressed them.

I am impressed by your ability to spot the `fits in one line` places.

I feel that sometimes `make style` can reformat in this case, but doesn't work well in a few places, as happened in this PR.

I dive a bit deeper, and found:

this one won't be reformatted by `make style`
```
    def prepare_inputs_for_generation(
        self,
        decoder_input_ids,
        past,
        use_cache=None,
        **kwargs,
    ):
```

but the following one will work well.

```
    def prepare_inputs_for_generation(
        self,
        decoder_input_ids,
        past,
        use_cache=None,
        **kwargs
    ):
```

The difference is the ending comma after the last argument. Is this a bug? I can open an issue if so.

File: src/transformers/generation_tf_utils.py
Comment (patrickvonplaten) [inline src/transformers/generation_tf_utils.py:685]: fine with me!

File: src/transformers/generation_tf_utils.py
Comment (patrickvonplaten) [inline src/transformers/generation_tf_utils.py:707]: that's a bit hacky - vision inputs should also work with `num_beams > 1` no? But ok for now until we do the big `generate` refactor @Rocketknight1

File: src/transformers/generation_utils.py
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:909]: can we remove this from the PR? This PR is for TF only no? I understand the need for this, but it's pretty hacky code that we should solve in a nicer way. I'll make the generate() function cleaner next week

File: src/transformers/models/deit/modeling_deit.py
Comment (patrickvonplaten) [inline src/transformers/models/deit/modeling_deit.py:480]: let's leave it for now - we can clean it up once the generate refactor is done: https://github.com/huggingface/transformers/issues/14421

File: src/transformers/models/encoder_decoder/modeling_tf_encoder_decoder.py
Comment (patrickvonplaten) [inline src/transformers/models/encoder_decoder/modeling_tf_encoder_decoder.py:156]: Thanks!

File: src/transformers/models/encoder_decoder/modeling_tf_encoder_decoder.py
Comment (patrickvonplaten) [reply src/transformers/models/encoder_decoder/modeling_tf_encoder_decoder.py:247]: Looks good to me!

File: tests/test_modeling_tf_vision_encoder_decoder.py
Comment (patrickvonplaten) [inline tests/test_modeling_tf_vision_encoder_decoder.py:2]:
```suggestion
# Copyright 2021- HuggingFace Inc. team.
```

File: tests/test_modeling_vision_encoder_decoder.py
Comment (patrickvonplaten) [inline tests/test_modeling_vision_encoder_decoder.py:67]: can we tackle all of this in a follow-up PR maybe? :-)

File: tests/test_modeling_tf_vision_encoder_decoder.py
Comment (patrickvonplaten) [inline tests/test_modeling_tf_vision_encoder_decoder.py:768]: Great tests! Could we add one more important test that shows that the architecture works as intended? 

E.g. could you simply copy paste your example here: https://huggingface.co/ydshieh/flax-vision-encoder-decoder-vit-gpt2-coco-en 

into a test so which a link to the image and an assert with the expected output. For this I think you need to upload a TF version of your model. 

It would be very nice to have it to be sure in the future that the model works correctly

**Review by patrickvonplaten — COMMENTED** [review]
This is a great addition @ydshieh! Thanks a lot for the contribution. 

From my side it would be great if we could:
- remove all pytorch specific changes that are not needed to get the TF version working (I'll tackle this in a future PR :-) )
- add one slow test that ensures that the model works correctly

Thanks a bunch!

**Comment by ydshieh** [comment]
> This is a great addition @ydshieh! Thanks a lot for the contribution.
> 
> From my side it would be great if we could:
> 
>     * remove all pytorch specific changes that are not needed to get the TF version working (I'll tackle this in a future PR :-) )
> 
>     * add one slow test that ensures that the model works correctly
> 
> 
> Thanks a bunch!

Hi @patrickvonplaten 

- OK for removing the changes on PT code.
- For the test, I think I forgot to do the same as I have done for PT/Flax

    https://github.com/huggingface/transformers/blob/8395f14de6068012787d83989c3627c3df6a252b/tests/test_modeling_vision_encoder_decoder.py#L666

    https://github.com/huggingface/transformers/blob/8395f14de6068012787d83989c3627c3df6a252b/tests/test_modeling_flax_vision_encoder_decoder.py#L467

    I will add it to TF test :)

File: src/transformers/generation_tf_utils.py
Comment (ydshieh) [reply src/transformers/generation_tf_utils.py:707]: Just as remark: The code inside this block treats `input_ids` as text-only inputs, for example `shape_list(input_ids)[-1]` assumes that the last dimension is the sequence dim.

For vision inputs, `generate` will be called only if it is a vision model as encoder in an encoder-decoder model (?). In this case, it is more the `decoder_input_ids` to be processed. And I think this is done in the next block

https://github.com/huggingface/transformers/blob/e2357599f7af1255b98729a3e0ef12bf2a081d44/src/transformers/generation_tf_utils.py#L722

It's not clear to me if an encoder-only vision model will need to call `generate()`. Maybe @NielsRogge can share some insights here (e.g. `ImageGPT`).

**Comment by ydshieh** [comment]
I removed the changes on PT code. I added 

```
class TFViT2GPT2ModelIntegrationTest(unittest.TestCase):
    @slow
    def test_inference_coco_en(self):
```
here
https://github.com/huggingface/transformers/blob/a179c4a5087e48127914e26e9895a455b42bac0d/tests/test_modeling_tf_vision_encoder_decoder.py#L775

**Comment by ydshieh** [comment]
Hi, @patrickvonplaten @Rocketknight1 @NielsRogge 

I removed the changes on PT code. I also added 

```
class TFViT2GPT2ModelIntegrationTest(unittest.TestCase):
    @slow
    def test_inference_coco_en(self):
```
here
https://github.com/huggingface/transformers/blob/b3255863540fac38d49e5067c14fef7ae66dc31f/tests/test_modeling_tf_vision_encoder_decoder.py#L769

This PR is ready for review when you have the time :-)

File: src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py
Comment (patrickvonplaten) [inline src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:310]: We should probably add this part: https://github.com/huggingface/transformers/blob/8f6373c61ca71d2d8176e181b587a2307d8358c8/src/transformers/models/encoder_decoder/modeling_tf_encoder_decoder.py#L402

here as well no? Otherwise it'll be difficult to load TF from PT

File: tests/test_modeling_tf_vision_encoder_decoder.py
Comment (patrickvonplaten) [inline tests/test_modeling_tf_vision_encoder_decoder.py:767]: Very nice!

File: src/transformers/generation_tf_utils.py
Comment (patrickvonplaten) [inline src/transformers/generation_tf_utils.py:632]: This whole function is in dire need of a refactor - I'll try to tackle this with @Rocketknight1 this month. Good for me the way it is not though

File: src/transformers/generation_tf_utils.py
Comment (patrickvonplaten) [inline src/transformers/generation_tf_utils.py:706]: nice!

**Review by patrickvonplaten — COMMENTED** [review]
This looks more or less ready to be merged to me:

- I think we should add the loading `from_pt` hack as well to `from_encoder_decoder(...)` here
- I think we can delete the file: docs/source/model_doc/auto.rst no?

File: src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py
Comment (ydshieh) [reply src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:310]: Yes, great catch! For some reason I missed it in this PR. I will check again this PR against `TFEncoderDecoderModel` code.

**Comment by ydshieh** [comment]
> * auto.rst no?

Yes. I added the (empty) file to commit by mistake during git rebase/merge.

File: src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py
Comment (ydshieh) [reply src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:311]: @patrickvonplaten  I corrected this part, and also updated the corresponding test (also forgot in the previous commits).

File: tests/test_modeling_tf_vision_encoder_decoder.py
Comment (ydshieh) [inline tests/test_modeling_tf_vision_encoder_decoder.py:703]: There was a `# TensorFlow => PyTorch` block, which did nothing (and if did, it would fail since we can't use `from_pretrained` along with `from_pt` or `from_tf` in this composite model.)

I removed it - need to remove the corresponding part in `test_modeling_tf_encoder_decoder.py` too in another PR.

File: src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py
Comment (patrickvonplaten) [reply src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:311]: Thanks a lot!

**Comment by patrickvonplaten** [comment]
@NielsRogge @Rocketknight1 - could you guys take a look here as well?

**Review by patrickvonplaten — APPROVED** [review]
Looks good to me now! I let @sgugger take a final look here

File: src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py
Comment (NielsRogge) [inline src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:94]: The PyTorch model (`VisionEncoderDecoderModel`) automatically creates the decoder_input_ids by shifting the labels. Is this not the case for the TF one?

File: src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py
Comment (NielsRogge) [inline src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:147]: Where does the _1 come from?

Comment (ydshieh) [reply src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:147]: I originally looked `TFRagModel` as reference for implementing the TF composite model, and saw

https://github.com/huggingface/transformers/blob/ac224bb0797c1ee6522d814139f3eb0a8947267b/tests/test_modeling_tf_rag.py#L973

I copied it and it works well (fixing the problems I had at that time), and I didn't think this part in more details.

I just spent some time on checking this again - and in fact, we can use `tf_vision_encoder_decoder_model` here, and `"tf_encoder_decoder_model` for `TFEncoderDecoderModel`.

Change this might break some user models - but since this is kind new models, and not popular yet, maybe it is worth the change. cc @patrickvonplaten , @Rocketknight1 , @sgugger, @LysandreJik for their thoughts on this.

(For `tf_rag_model_1`, maybe there is particular reason to make something work though.)

File: src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py
Comment (ydshieh) [reply src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:94]: No. I started this work before the new change (about `decoder_input_ids`) was done in `VisionEncoderDecoderModel`, and didn't follow it.

Would it be possible to leave it to another PR - I can make it, but prefer in a separate PR.

Comment (NielsRogge) [reply src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:94]: Ok makes sense!

Comment (NielsRogge) [reply src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:94]: See #14469

File: src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py
Comment (ydshieh) [reply src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py]: I changed to "tf_vision_encoder_decoder_model" - better not to continue with the strange `_1`

File: src/transformers/models/encoder_decoder/modeling_tf_encoder_decoder.py
Comment (sgugger) [inline src/transformers/models/encoder_decoder/modeling_tf_encoder_decoder.py:154]: Those two lines need fixing
```suggestion
   [`~TFAutoModel.from_pretrained`] class method for the encoder and
   [`~TFAutoModelForCausalLM.from_pretrained`] class method for the decoder.
```

File: src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py
Comment (sgugger) [inline src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:133]:
```suggestion
            - With a *decoder_* prefix which will be input as `**decoder_kwargs` for the decoder forward function.
```

File: src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py
Comment (sgugger) [inline src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:143]:
```suggestion
    decoder when created with the [`~TFAutoModel.from_pretrained`] class method for the encoder and
    [`~TFAutoModelForCausalLM.from_pretrained`] class method for the decoder.
```

File: src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py
Comment (sgugger) [inline src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:258]:
```suggestion
        Initializing `TFVisionEncoderDecoderModel` from a pytorch checkpoint is not supported currently.
```

File: src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py
Comment (sgugger) [inline src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:324]:
```suggestion
            encoder_pretrained_model_name_or_path (`str`, *optional*):
```

File: src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py
Comment (sgugger) [inline src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:334]:
```suggestion
            decoder_pretrained_model_name_or_path (`str`, *optional*, defaults to *None*):
```

**Review by sgugger — APPROVED** [review]
There are a few typos in the docstrings, but LGTM once it's fixed!
Thanks for all your work on this @ydshieh !

**Comment by ydshieh** [comment]
Applied @sgugger  review suggestions. Failed tests are unrelated.

**Comment by sgugger** [comment]
Thanks again for all your work on this!

**Pull request merged**
